### DISCUSSION TITLE
Semantically mark multiline prompt to ignore line continuation marks '...'

### DISCUSSION THREAD

**Discussion by eugenesvk** [wezterm/wezterm#3130] [Q&A]
Let's say we have a shell that supports fully editable multiline prompts and to differentiate between a new line within a multiline prompt and a regular newline prepends to each continuation line a few dots

For example, this code...
```py
if True:
  echo 1
  echo 2
```

... in a shell is shows like this:
```xsh
if True:
..  echo 1
..  echo 2
```

What I don't understand is how to add [semantic prompt](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md) markers that would allow selection of all the code without the `..` marks

If I addadd prompt_start/prompt_end_input_start marks before/after `..` like so:
`OSC133;P;k=c\007` `..` `OSC133;B\007`
```
↑ start of continuation prompt
                 ↑ continuation marks
                    ↑ end of prompt and start of user input, terminated by a OSC "133;C" or another prompt (OSC "133;P")
```
then I can multiclick to select `echo 1` instead of `..  echo 1` because input `echo 1` is properly marked

But then I can select all the lines since (I guess) the end of prompt marks break continuation :(
But then if I don't add the end-of-prompt marks, I can selec the whole multiline area, but with the `..` continuation marks :(

(is this related to this issue in the semantic prompt docs, meaning that what I'd like to do is impossible since it's not implemented yet?
> Issue: Possibly add a way to mark indentation as distinct from prompts.

)

I've also noticed that if I add a start-of-continuation-prompt mark after each `..`, then I lose the ability to select the whole input area, the first `if True:` line is lost, the selection starts at `  echo 1`  (but at least the first `..` are excluded)

**Comment by wez** [comment]
This is a limitation of the `SemanticType` enum:
https://github.com/wez/wezterm/blob/main/termwiz/src/cell.rs#L168
which doesn't understand about continuation (or right) prompts.

To extend it to support that, I think the storage for the semantic zone needs to move out of CellAttributes:
https://github.com/wez/wezterm/blob/main/termwiz/src/cell.rs#L277

and into the `FatAttributes` struct:
https://github.com/wez/wezterm/blob/main/termwiz/src/cell.rs#L80

so that it can then be modified to hold more information.

Then the logic in `term/src/terminalstate/performer.rs` that interprets and applies the escape sequences needs to be updated to pass down the additional information.

With the data in place, it would then be possible to teach the semantic zone aware selection logic how to traverse the continuation zones.

**Reply by eugenesvk** [reply]
Oh, I see, so it's more of a limitation of the WezTerm itself
(I think you already have a broader semantic zone issue https://github.com/wez/wezterm/issues/115, so I'll just mention it there)
